### ISSUE TITLE
publish_address should default to bound_address

### ISSUE THREAD

**Issue by aleszoulek** [elastic/elasticsearch#14561]
When `network.publish_host` is not specified in a configuration, ES should try to reuse the value set for `network.bind_host`.

The current behaviour is that for following configration
```
....
network.bind_host: 10.88.12.218
# network.publish_host: NOT SET
```

The ES node will bind to correct network interface, but it will publish a different one (not sure how it's currently determined - ?first non-loopback inet? ). Which doesn't seem to make much sense.

Relevant log line:
```
[2015-11-05 14:47:52,744][INFO ][transport                ] [...] bound_address {inet[/10.88.12.218:9300]}, publish_address {inet[/10.88.64.171:9300]}
```

Workround is easy. Just set both `bind_host` and `publish_host` to a same value.

**Comment by rmuir** [comment]
This stuff is fixed in 2.0. basically the code in 1.x is geared around the default of binding to `0.0.0.0`, so that logic there for publish address "works in most cases".

In 2.0, if you don't set publish address, we look at the actual addresses bound, sort them by ipv4/ipv6 stack preference, then by reachability scope, and select the first one.

**Comment by nik9000** [comment]
This is indeed confusing. You are supposed to set `network.host` 99% of the time and ignore the other two. But the docs aren't great there.

**Comment by rmuir** [comment]
> This is indeed confusing. You are supposed to set network.host 99% of the time and ignore the other two. But the docs aren't great there.

Agreed, besides the problems in 1.x, I think its good for publish host selection to check bind_host before host in case the user did just this. There is no need to default publishing to host directly, to something we potentially didnt bind to, unless someone explicitly set publish host.

**Comment by aleszoulek** [comment]
That's my fault indeed. I should have be more careful when reading an example configuration. I've seen the "Network And HTTP" section and the first option that's there is `#network.bind_host:` and this is what I've set. If I'd read the rest of the config file, there's just few lines bellow:
```
# Set both 'bind_host' and 'publish_host':
# 
# network.host: 192.168.0.1
```

Although it would be nice if it did what @rmuir described (check bind_host in publish host selection).

**Comment by rmuir** [comment]
Yeah, if its easy to do that check, we should do it. anything we can do to make it less confusing. 

The documentation really needs improvement here (we've seen other problems). Basically network.host should be the recommended way, and the bind/publish stuff should be buried to the bottom of the page in some kind of advanced options section: you shouldn't mess with that unless you have a more complex configuration (like proxies, static NAT, something like that)

**Comment by clintongormley** [comment]
Closed by #15360
